Update nest integration with fixes from initial PR

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Address feedback from the original PR #41689, namely:
- Address type in should_poll
- Remove unnecessary async_update code
- Throw explicit exceptions in config flow for "shouldn't happen" cases
- Simplify create_task logic

Also adds test coverage for device information, fixing tests to use
the correct type name and simplifying the type lookup.



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Update nest integration with fixes from initial PR

Address feedback from the original PR #41689, namely:
- Address type in should_poll
- Remove unnecessary async_update code
- Throw explicit exceptions in config flow for "shouldn't happen" cases
- Simplify create_task logic

Also adds test coverage for device information, fixing tests to use
the correct type name and simplifying the type lookup.

MartinHjelmare commented on homeassistant/components/nest/sensor_sdm.py:
```diff
@@ -89,28 +96,17 @@ def device_model(self):
-        if self._device.type == "sdm.devices.types.THERMOSTAT":
-            return "Thermostat"
-        return None
+        return DEVICE_TYPE_MAP.get(self._device.type, None)
```
`None` is the default value returned if the key is missing in the dict.

MartinHjelmare commented on homeassistant/components/nest/sensor_sdm.py:
```diff
@@ -89,28 +96,17 @@ def device_model(self):
         self.async_on_remove(
-            async_dispatcher_connect(self.hass, SIGNAL_NEST_UPDATE, async_update_state)
+            async_dispatcher_connect(
+                self.hass, SIGNAL_NEST_UPDATE, self.async_update_ha_state
```
I think we can replace `self.async_update_ha_state` with `self.async_write_ha_state`.

allenporter replied:
When I replace that, the tests fail because the update is not reflected in state.   I think the only difference i see is the "async" keyword prefix.  I'm fairly new to python async so i'm missing the answer

MartinHjelmare replied:
It should work. Maybe something else in the tests isn't set up correctly?

balloob replied:
fixed it in a404a220b3

balloob replied:
(and then I had a whoopsie and reverted the change for legacy in 6b44456)

allenporter replied:
> In the fake subscriber in the tests we have decorated a coroutine function `async_device_manager` as a property. I dont think that is correct.

Thanks, I'll fix this in the underlying library, then update HA and tests.

MartinHjelmare commented on homeassistant/components/nest/config_flow.py:
```diff
@@ -111,7 +111,7 @@ async def async_step_user(self, user_input=None):
         """Handle a flow start."""
         if self.is_sdm_api():
-            return None
+            raise InvalidStateError("Step only supported for legacy API")
```
I'm not sure if we should use this exception. Until now it is only used by the core StateMachine. I'd advise to define a new exception in this module that inherits HomeAssistantError to raise.

allenporter committed:
Update nest integration review feedback

- Add a test for unknown types, and simplify the code
- Define an exception for unexpected states, avoiding use of existing InvalidStateErorr that is only used by the state
  machine.

allenporter committed:
Update nest exception inheritence

allenporter committed:
Update nest async callbacks

Replace self.async_update_ha_state with call to self.async_write_ha_state.

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
In the fake subscriber in the tests we have decorated a coroutine function `async_device_manager` as a property. I dont think that is correct.

MartinHjelmare commented:
We should also move the block till done inside the patch context manager block when setting up the integration.

balloob commented on homeassistant/components/nest/__init__.py:
```diff
@@ -176,7 +180,7 @@ def handle_event(self, event_message: EventMessage):
         # properties which the DeviceManager should already have received.
         # Send a signal to refresh state of all listening devices.
-        dispatcher_send(self._hass, SIGNAL_NEST_UPDATE)
+        async_dispatcher_send(self._hass, SIGNAL_NEST_UPDATE)
```
Background on this change: the SDM integration runs inside the event loop, and so do it's callbacks. So the callbacks should use the async versions of functions.

balloob replied:
Tagged this PR for 117 because of this change.

allenporter commented:
I just ran https://developers.home-assistant.io/docs/development_catching_up/ to catch up and worry i may have undone those extra commits somehow.

allenporter committed:
Fix nest dispatcher sync method

Update the dispatcher_send call to async_dispatcher_send because
the integration and its callbacks run inside the event loop, so
therefore they must use the async versions of the functions.

Replace use of async_update_ha_state with async_write_ha_state since
no update call is needed, which now bases with the above fix.

allenporter commented:
> I just ran https://developers.home-assistant.io/docs/development_catching_up/ to catch up and worry i may have undone those extra commits somehow.

I've re-added the changes from @balloob manually since i messed up that rebase / push.  This should be ready to go now.  I'll followup with the property change with a new release of python-google-nest-sdm.

MartinHjelmare commented:
Can we move the block till done in the test now?

allenporter committed:
Move block until done inside the patch block

Address PR feedback to move setup inside the patch context manager.

allenporter commented:
> Can we move the block till done in the test now?

Thanks, missed that comment.  Done.
